### PR TITLE
[EraVM][AsmParser] Migrate to new asm syntax

### LINKED ISSUE #441: [MC][AsmParser] Rename mnemonics according to the specification
At now, most instructions are supported with _some_ mnemonic, but these mnemonics can differ in the public specification, pre-existing code (LLVM backend and standalone assembler) and possibly the updated spec.

This issue covers instruction mnemonics in a broad sense (for example, the `<op>` part in `context.<op>`) - those cases that can be resolved by simply changing string literals in C++ code or `*.td` files without changing any logic.

See #440 for conditional codes.

### PR DESCRIPTION
**Pull request by atrosinenko**
Resolves #441.

### CODE REVIEW

**Comment by atrosinenko** [comment]
At now, the changes are split into multiple commits for clarity.

Please note that while these changes are expected to be complete, they are not expected to pass the CI until switching to the new assembler.

**Comment by asl** [comment]
@sayon @hedgar2017 Will you please check if the resulting syntax is what you've been thinking of?

**Comment by hedgar2017** [comment]
> @sayon @hedgar2017 Will you please check if the resulting syntax is what you've been thinking of?

Hey @asl, we're doing the final review of modifiers and will share a short list of changes soon.

**Comment by hedgar2017** [comment]
@asl could you point to the specific place I should review?

**Comment by hedgar2017** [comment]
@asl @atrosinenko could you kindly remove the `{...}` syntax for modifiers while we're discussing final detail?

Sorry for telling this late. Overlooked it before, and now we're sure this syntax is redundant.

**Comment by asl** [comment]
@hedgar2017 The files in [llvm/test/MC/EraVM/asm-parser/](https://github.com/matter-labs/era-compiler-llvm/pull/576/files#diff-f2451450a53718cb8af9eb45ad874febac6fde4dda2049f46e05b1db4f53acac) essentially show diff before / after (note that [arith-instructions-errors.s](https://github.com/matter-labs/era-compiler-llvm/pull/576/files#diff-45a1e54797770d56ca9a328d49c493a63a6c8c2df2e4133573923ae609fb7835) checks for errors, so the syntax there is not correct for a purpose :) ).

File: llvm/lib/Target/EraVM/EraVMOpcodes.td
Comment (hedgar2017) [inline llvm/lib/Target/EraVM/EraVMOpcodes.td]: Could we capitalize `l`, and make it `logL1`?
`l` is hardly readable, and Layer 1 is usually spelled `L1`.

File: llvm/lib/Target/EraVM/EraVMOpcodes.td
Comment (hedgar2017) [inline llvm/lib/Target/EraVM/EraVMOpcodes.td]: Let's change `f` => `i`, `is_first` => `is_initializer`.
`i` (initializer) better reflects the purpose of this instruction.

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
Hey @asl @atrosinenko, it addition to my comments above, I'm curious if we can implement this kind of printing/formatting?
```
add   az          r1, r2, r3
add   az!         r1, r2, r3
sub   z.s!        r1, r2, r3
```
We have formatting and readability issues compared to formats of generic-purpose architectures and EVM.
Could we move modifiers to another column? This way we could allocate ~8 symbols for mnemonics, ~12 for modifiers, and arguments won't be jumping across the entire screen.

cc @akiramenai

**Comment by atrosinenko** [comment]
> could you kindly remove the {...} syntax for modifiers while we're discussing final detail?

@hedgar2017 Do you mean replacing the syntax like this
```
  callf.{st,sh}           r3, r4, @handler
```
with
```
  callf.st.sh           r3, r4, @handler
```
In that case, should `callf.sh.st` be supported as well?

**Comment by hedgar2017** [comment]
> > could you kindly remove the {...} syntax for modifiers while we're discussing final detail?
> 
> @hedgar2017 Do you mean replacing the syntax like this
> 
> ```
>   callf.{st,sh}           r3, r4, @handler
> ```
> 
> with
> 
> ```
>   callf.st.sh           r3, r4, @handler
> ```
> 
> In that case, should `callf.sh.st` be supported as well?

Yes, order remains arbitrary, but let's resolve [this first](https://github.com/matter-labs/era-compiler-llvm/pull/576#pullrequestreview-2083397147).

**Comment by asl** [comment]
> Could we move modifiers to another column? This way we could allocate ~8 symbols for mnemonics, ~12 for modifiers, and arguments won't be jumping across the entire screen.

This reminds me of FORTAN and punchcards :)

This is a bit non-trivial. Semantically in the spec you're having "instruction" and "modifiers". However, from the behavior / codegen perspective all these are separate instructions. So, `log` and `log.f` are separate instructions. We would need to change the assembler parser, as this requires merging the tokens together (normally tokens are separated by whitespace) during the parsing. Certainly doable, but would require some changes with some of those being a bit fundamental in how assembler parser is implemented.

**Comment by hedgar2017** [comment]
@asl understood :) So let's keep them together (`mnemonic.mod1.mod2...modN[!]`).
Could we estimate the maximum length of all modifiers and provide it to the printer, e.g. 5+5*3 if we can use 5 mnemonics at most?
Parser will accept any whitespace, but in printer/debugger we'll see proper offsets. (not sure where the debugger will be receiving the formatting from)

**Comment by asl** [comment]
> Could we estimate the maximum length of all modifiers and provide it to the printer, e.g. 5+5*3 if we can use 5 mnemonics at most?

Printer is streaming. So, it does not see the complete code. Functions are printed independently and basic blocks are also printed independently. So, we'd need to make a complete sweep over the code before printing to estimate this (and have some side information about printer instruction lengths...). Note that bunch of stuff is auto-generated (with some sane assumptions), so this might break this part.

Anyway, @atrosinenko will try to experiment and see what will be easier / more straightforward to implement

**Comment by hedgar2017** [comment]
@asl, nah, not during printing. Just from the spec.
@sayon what could be the longest string of mnemonic+modifiers? :)

**Comment by hedgar2017** [comment]
Anyway, we can experiment with offsets in another PR.

**Comment by asl** [comment]
> Anyway, we can experiment with offsets in another PR.

Currently we are emitting e.g.:
```
\tadd\tr1, r2, r3
```

We will double check if there is always a tab between mnemonic and operand. But I guess, tab width of 8 is not enough? Or maybe you're having it 4 locally?

**Comment by sayon** [comment]
@hedgar2017 i would say `callf.st.sh.nz` should be the longest possible instruction. 13 chars, so we could force the first column to be 13 chars wide. This is passable.


```
mul            r1, r2, r3, r4
sub.s!         r1, r2, r3
add            r1, r2, r3
callf.st.sh.nz r1, something, something
add            r1, r2, r3
sub            r3, r2, r1
```

**Comment by hedgar2017** [comment]
> > Anyway, we can experiment with offsets in another PR.
> 
> 
> 
> Currently we are emitting e.g.:
> 
> ```
> 
> \tadd\tr1, r2, r3
> 
> ```
> 
> 
> 
> We will double check if there is always a tab between mnemonic and operand. But I guess, tab width of 8 is not enough? Or maybe you're having it 4 locally?

Given @sayon's advice, we would be good to go with 2 tabs, but this will break if the editor setting is not 8.

**Comment by atrosinenko** [comment]
> Given sayon's advice, we would be good to go with 2 tabs, but this will break if the editor setting is not 8.

@hedgar2017 That is, keep `\t{mnemonic}\t{operands}` for now and adjust it further via a separate PR, if needed? Or do you propose adding one more literal tab after the mnemonic, such as `\tadd\t\tr1, r2, r3`?

**Comment by hedgar2017** [comment]
> > Given sayon's advice, we would be good to go with 2 tabs, but this will break if the editor setting is not 8.
> 
> 
> 
> @hedgar2017 That is, keep `\t{mnemonic}\t{operands}` for now and adjust it further via a separate PR, if needed? Or do you propose adding one more literal tab after the mnemonic, such as `\tadd\t\tr1, r2, r3`?

Let's play with it in a separate PR.
I'd like us to switch to spaces because I have been through a hell of different tab settings in various IDEs.
AFAIK most of modern styles are switching to spaces nowadays.

File: llvm/lib/Target/EraVM/EraVMOpcodes.td
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/EraVMOpcodes.td]: Fixed.

File: llvm/lib/Target/EraVM/EraVMOpcodes.td
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/EraVMOpcodes.td]: Updated here and `log.f` below.

**Comment by atrosinenko** [comment]
I pushed a rename of `log*` instructions. Considering removing `{...}` from far call's modifiers, I would like to update it via separate PR and order it **before** this one, so that as much changes as possible are checked by a full set of CI tests.

I plan not rebasing this PR every time, but merge to `main` as much as possible first, and then rebase this PR once (and squash its commits, I guess).

File: llvm/lib/Target/EraVM/EraVMOpcodes.td
Comment (hedgar2017) [inline llvm/lib/Target/EraVM/EraVMOpcodes.td]: Change the def as well please.

**Review by hedgar2017 — CHANGES_REQUESTED** [review]

**Comment by atrosinenko** [comment]
Well, updated the dependency PR and had to rebase with manual conflict resolution 30 minutes after telling I don't plan to rebase this PR often... Updated the tests for `log*` instructions as well.

File: llvm/lib/Target/EraVM/EraVMOpcodes.td
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/EraVMOpcodes.td]: Changed `First` to `Initial`.

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
We're waiting for your PR removing `{...}` syntax, then rebasing and merging this one, correct @atrosinenko ?

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (atrosinenko) [inline llvm/lib/Target/EraVM/EraVMInstrInfo.td]: @asl `r` suffixes can be dropped from `def` names of a few instructions here, I guess (such as `L1Initialr -> L1Initial` or `PCOMPr -> PCOMP`). There is only one variant of each instruction and the suffixes should be `rr` or `rrr` anyway.

**Comment by atrosinenko** [comment]
> We're waiting for your PR removing {...} syntax, then rebasing and merging this one, correct @atrosinenko ?

@hedgar2017 In addition, there are two dependency PRs: #571 (approved, will merge it after CI finishes) and #575 (trivial, on review).

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/EraVMInstrInfo.td]: (this can be updated later, of course)

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
LGTM!
Could you turn conversations not addressed to anyone particular into source code comments?

**Comment by atrosinenko** [comment]
Pushed a few last-minute fixes for parsing `incsp`, `decsp` and for `logL1`-related error messages in tests.

Considering `logL1`, it turned out that `EraVMAsmParser::ParseInstruction` method receives `Name` in lower case. That means, both `logL1` and `logl1` are accepted, but `logL1` is always printed:
```
$ ./bin/llvm-mc -arch=eravm <<< 'logL1 r1, r2'
        .text
        logL1   r1, r2
$ ./bin/llvm-mc -arch=eravm <<< 'logl1 r1, r2'
        .text
        logL1   r1, r2
```

When complaining about unknown instruction mnemonic, the mnemonic is always printed in lower case in error message (but not in cited source file):
```
$ ./bin/llvm-mc -arch=eravm <<< 'logL1.i.i r1, r2'
        .text
<stdin>:1:1: error: unknown mnemonic: logl1.i.i
logL1.i.i r1, r2
^
$ ./bin/llvm-mc -arch=eravm <<< 'AbCdEffg r1, r2'
        .text
<stdin>:1:1: error: unknown mnemonic: abcdeffg
AbCdEffg r1, r2
^
```

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/EraVMInstrInfo.td]: Created a separate follow-up PR to move this discussion there: #583.

**Comment by atrosinenko** [comment]
No functional changes, added a few more test cases: test that `ldm(i)` and `stm(i)` do not have a "default" modifier.

**Review by hedgar2017 — DISMISSED** [review]

**Comment by PavelKopyl** [comment]
This change was included into #632.

**Pull request closed without merging**
